examples/iso_dh/DY.Example.DH.Protocol.Total.fst

```diff
@@ -0,0 +1,156 @@
+
+[@@ with_bytes bytes]
+type message3 = {
+  sg:bytes;
+}
+
+%splice [ps_message3] (gen_parser (`message3))
+%splice [ps_message3_is_well_formed] (gen_is_well_formed_lemma (`message3))
+
+
+[@@ with_bytes bytes]
+type message =
+  | Msg1: msg:message1 -> message
+  | Msg2: msg:message2 -> message
+  | Msg3: msg:message3 -> message
+
+%splice [ps_message] (gen_parser (`message))
+%splice [ps_message_is_well_formed] (gen_is_well_formed_lemma (`message))
+
+instance parseable_serializeable_message: parseable_serializeable bytes message = mk_parseable_serializeable ps_message
+
+// Definition of signature terms
+[@@ with_bytes bytes]
+type sig_message2 = {
+  a:principal;
+  gx:bytes;
+  gy:bytes;
+}
+
+%splice [ps_sig_message2] (gen_parser (`sig_message2))
+%splice [ps_sig_message2_is_well_formed] (gen_is_well_formed_lemma (`sig_message2))
+
+[@@ with_bytes bytes]
+type sig_message3 = {
+  b:principal;
+  gx:bytes;
+  gy:bytes;
+}
+
+%splice [ps_sig_message3] (gen_parser (`sig_message3))
+%splice [ps_sig_message3_is_well_formed] (gen_is_well_formed_lemma (`sig_message3))
+
+[@@ with_bytes bytes]
+type sig_message =
+  | SigMsg2: msg:sig_message2 -> sig_message
+  | SigMsg3: msg:sig_message3 -> sig_message
+
+%splice [ps_sig_message] (gen_parser (`sig_message))
+%splice [ps_sig_message_is_well_formed] (gen_is_well_formed_lemma (`sig_message))
+
+instance parseable_serializeable_sig_message: parseable_serializeable bytes sig_message = mk_parseable_serializeable ps_sig_message
+
+
+(*** Message Processing ***)
+
+// Alice generates message 1
+val compute_message1: principal -> bytes -> bytes
+let compute_message1 alice x =
+  let gx = dh_pk x in
+  let msg = Msg1 {alice; gx} in
+  serialize message msg
+
+// Bob parses message 1
+val decode_message1: bytes -> option message1
+let decode_message1 msg1_bytes =
+  let? msg1 = parse message msg1_bytes in
+  // These lines are the...
+  guard (Msg1? msg1);?
+  Some (Msg1?.msg msg1)
+  // ...short version of the following match:
+  (*    
+  match msg1 with
+  | Msg1 msg -> Some msg
+  | _ -> None
+  *)
+
+// Bob generates message 2
+val compute_message2: principal -> principal -> bytes -> bytes -> bytes -> bytes -> bytes
+let compute_message2 alice bob gx gy sk_b n_sig =
+  let sig_msg = SigMsg2 {a=alice; gx; gy} in
+  let sg = sign sk_b n_sig (serialize sig_message sig_msg) in
+  let msg = Msg2 {bob; gy; sg} in
+  serialize message msg
+
+// Alice parses message 2
+val decode_message2: bytes -> principal -> bytes -> bytes -> option message2
+let decode_message2 msg2_bytes alice gx pk_b =
+  let? msg2_parsed = parse message msg2_bytes in
+  guard (Msg2? msg2_parsed);?
+  let msg2 = Msg2?.msg msg2_parsed in
+  // Verify the signature contained in the message 2
+  // with the gy value from the message and the gx
+  // value from Alice's state.
+  let gy = msg2.gy in
+  let sig_msg = SigMsg2 {a=alice; gx; gy} in
+  if verify pk_b (serialize sig_message sig_msg) msg2.sg then Some msg2
+  else None
```

Comment:
I also think it looks better but it makes the code harder to understand for people from outside. So I would keep it like this, but if you want to have the ``guard`` function here, I can change it and make a comment.

Reply:
As you wish, I think using `guard` is better for the following reasons:
- it is a [standard monadic function](https://hackage.haskell.org/package/base-4.16.3.0/docs/Control-Monad.html#v:guard) (at least in haskell)
- with this style the code is more future-proof if you want to add other monadic computations after the signature verification (unless you want to nest code in the `else` branch, but this is the goal of `guard` to avoid that)
- it makes the code style more consistent because there are other `guard`s in other functions in the same file